src/main/java/com/mycompany/physicstoolbox/Viewport.java

```diff
@@ -144,6 +168,29 @@ private void fallOnePixel(int gridX, int gridY) {
             grid[gridY + 1][gridX] = temp;
         }
         
+        // Randomly determines a direction to flow based on the current pixel's available choices
+        private void flowOnePixel(int gridX, int gridY) {
+            boolean leftOccupied = gridX - 1 < 0 || grid[gridY][gridX - 1].containsSubstance();
+            boolean rightOccupied = gridX + 1 >= grid[gridY].length || grid[gridY][gridX + 1].containsSubstance();
+            
+            // -1 denotes left, 1 denotes right
+            int flowDirection;
+            if(leftOccupied && rightOccupied) {
+                return;
+            } else if(!leftOccupied && rightOccupied) {
+                flowDirection = -1;
+            } else if(leftOccupied && !rightOccupied) {
+                flowDirection = 1;
+            } else {
+                flowDirection = !rand.nextBoolean() ? -1 : 1;
+            }
+            
+            // Applying the corresponding flow
+            Pixel temp = grid[gridY][gridX];
+            grid[gridY][gridX] = grid[gridY][gridX + flowDirection];
+            grid[gridY][gridX + flowDirection] = temp;
+        }
```

Comment:
Not really random, it just flows in the available grid location. Am I missing something?